lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacetLabels.java

```diff
@@ -0,0 +1,184 @@
+     * @throws IOException when a low-level IO issue occurs
+     */
+    public FacetLabel nextFacetLabel(int docId) throws IOException {
+      if (currentDocId != docId) {
+        assert docId > currentDocId: "docs out of order!  currentDocId=" + currentDocId + " docId=" + docId;
+        ordinalsSegmentReader.get(docId, decodedOrds);
+        currentDocId = docId;
+        currentPos = decodedOrds.offset;
+      }
+
+      int endPos = decodedOrds.offset + decodedOrds.length;
+      assert currentPos <= endPos;
+
+      if (currentPos == endPos) {
+        // no more FacetLabels
+        return null;
+      }
+
+      int ord = decodedOrds.ints[currentPos++];
+      return taxoReader.getPath(ord);
+    }
+
+    private boolean isDescendant(int ord, int ancestorOrd) {
+      while (ord != INVALID_ORDINAL && ord != ROOT_ORDINAL) {
+        if (parents[ord] == ancestorOrd) {
+          return true;
+        }
+        ord = parents[ord];
+      }
+      return false;
+    }
+
+    /**
+     * Retrieves the next {@link FacetLabel} for the specified {@code docId} under the requested {@code facetDimension},
+     * or {@code null} if there are no more. This method has state: if the provided {@code docId} is the same as the
+     * previous invocation, it returns the next {@link FacetLabel} for that document.  Otherwise, it advances to
+     * the new {@code docId} and provides the first {@link FacetLabel} for that document, or {@code null} if that document
+     * has no indexed facets.  Each new {@code docId} must be in strictly monotonic (increasing) order.
+     *
+     * <p><b>NOTE</b>: this method loads the {@code int[] parents} array from the taxonomy index</p>
+     *
+     * @param docId input docId provided in non-decreasing order
+     * @return the first or next {@link FacetLabel}, or {@code null} if there are no more
+     * @throws IOException if {@link TaxonomyReader} has problems getting path for an ordinal
+     */
+    public FacetLabel nextFacetLabel(int docId, String facetDimension) throws IOException {
+      final int parentOrd = taxoReader.getOrdinal(new FacetLabel(facetDimension));
+      assert parentOrd != INVALID_ORDINAL : "Category ordinal not found for facet dimension: " + facetDimension;
+
+      if (currentDocId != docId) {
+        assert docId > currentDocId: "docs out of order!  currentDocId=" + currentDocId + " docId=" + docId;
+        ordinalsSegmentReader.get(docId, decodedOrds);
+        currentPos = decodedOrds.offset;
+        currentDocId = docId;
+      }
+
+      if (parents == null) {
+        parents = taxoReader.getParallelTaxonomyArrays().parents();
+      }
+
+      int endPos = decodedOrds.offset + decodedOrds.length;
+      assert currentPos <= endPos;
+
+      for (; currentPos < endPos; ) {
+        int ord = decodedOrds.ints[currentPos++];
+        if (isDescendant(ord, parentOrd) == true) {
+          return taxoReader.getPath(ord);
```

Comment:
Hmm when we switch to `BinaryDocValues` for this lookup, we could fix this class to persistently hold a single `BinaryDocValues`, and improve performance for caller that need to look up multiple ordinals, which is the common case.

Reply:
By this class you mean `TaxonomyReader` because that is where `BinaryDocValues` will be cached as part of [LUCENE-9476](https://issues.apache.org/jira/browse/LUCENE-9476) improving performance for callers that need to lookup multiple ordinals ?

Reply:
Hmm, I was thinking this new per-segment class (`FacetLabelReader`) would hold onto a single `BinaryDocValues` instance, but you're right, that won't work.

The facet ordinals are stored / compacted in sorted order in a single `BinaryDocValues` field which we decode here into `decodedOrds`.  Resolving those ordinals to `FacetLabel` is costly, currently retrieving one stored document per ord.  After LUCENE-9450 (switching from stored fields to BDV), we could pull a new BDV for each unique `docId` passed to `nextFacetLabel`, and then bulk resolve all the ordinals?

Anyway, we can pursue all of this later -- this PR already looks awesome -- progress not perfection!

Reply:
The new per-segment class (`FacetLabelReader`) holds an instance of `DocValuesOrdinalsReader` which internally holds a reusable instance of `BinaryDocValues`.

In the `nextFacetLabel()` method `ordinalsSegmentReader.get(docId, decodedOrds)` that does the actual decoding of the ordinals is only called **once** if the input docId is the same as the one supplied in prior invocation of `nextFacetLabel()`.  

Subsequent invocations fetch the ordinals from `FacetLabelReader.decodedOrds` and use the ordinal to lookup FacetLabels using `taxoReader.getPath(ord)`. 

Yes, after LUCENE-9450, we can bulk resolve all the ordinals using a different API, something like -> `List<FacetLabels> taxoReader.getPaths(IntRef decodedOrds)` and add a new API - `FacetLabelReader.allFacetLabels(int docId)`

Reply:
Yeah, or we could bulk map all ord -> FacetLabel as soon as caller goes to the next `docId`, and then iterate them one by one with this API.

Anyway, we don't need to solve that here.